apm-agent-plugins/apm-spring-webmvc-plugin/src/test/java/co/elastic/apm/agent/spring/webmvc/DispatcherServletRenderInstrumentationTest.java

```diff
@@ -0,0 +1,103 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+package co.elastic.apm.agent.spring.webmvc;
+
+import co.elastic.apm.agent.MockReporter;
+import co.elastic.apm.agent.bci.ElasticApmAgent;
+import co.elastic.apm.agent.configuration.SpyConfiguration;
+import co.elastic.apm.agent.impl.ElasticApmTracer;
+import co.elastic.apm.agent.impl.ElasticApmTracerBuilder;
+import co.elastic.apm.agent.servlet.ServletInstrumentation;
+import net.bytebuddy.agent.ByteBuddyAgent;
+import org.junit.AfterClass;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.junit.jupiter.api.AfterAll;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.BeforeEach;
+import org.springframework.stereotype.Controller;
+import org.springframework.test.web.servlet.MockMvc;
+import org.springframework.test.web.servlet.setup.MockMvcBuilders;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.servlet.ViewResolver;
+import org.springframework.web.servlet.view.InternalResourceViewResolver;
+
+import java.util.Arrays;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
+
+public class DispatcherServletRenderInstrumentationTest {
+
+    private static MockReporter reporter;
+    private static ElasticApmTracer tracer;
+    private MockMvc mockMvc;
+
+    @BeforeClass
+    @BeforeAll
+    public static void setUpAll() {
+        reporter = new MockReporter();
+        tracer = new ElasticApmTracerBuilder()
+            .configurationRegistry(SpyConfiguration.createSpyConfig())
+            .reporter(reporter)
+            .build();
+        ElasticApmAgent.initInstrumentation(tracer, ByteBuddyAgent.install(),
+            Arrays.asList(new ServletInstrumentation(), new DispatcherServletRenderInstrumentation()));
+    }
+
+    @AfterClass
+    @AfterAll
+    public static void afterAll() {
+        ElasticApmAgent.reset();
+    }
+
+    @Before
+    @BeforeEach
+    public void setup() {
+        this.mockMvc = MockMvcBuilders
+            .standaloneSetup(new MessageController())
+            .setViewResolvers(jspViewResolver())
+            .build();
+    }
+
+    private ViewResolver jspViewResolver() {
+        InternalResourceViewResolver bean = new InternalResourceViewResolver();
+        bean.setPrefix("/static/");
+        bean.setSuffix(".jsp");
+        return bean;
+    }
+
+    @Controller
+    public static class MessageController {
+        @GetMapping("/test")
+        public String test() {
+            return "message-view";
+        }
+    }
+
+    @Test
+    public void testCallOfDispatcherServletRender() throws Exception {
+        this.mockMvc.perform(get("/test"));
+        assertEquals(1, reporter.getTransactions().size());
+        assertEquals(2, reporter.getSpans().size());
```

Comment:
what's the other span?

Reply:
i will add check on that span isAlreadyMonitored.

Reply:
So what is it? If it's not related to this test (e.g. from the setup step or something) - just start the test with `reporter.reset();` to start clean. Otherwise, if it's something we don't want to report we shouldn't.

Reply:
it is span of DispatcherServlet#render. I don't understand why it's appear twice on test.

Reply:
Can you debug that? Because we wouldn't want to send twice if only one is valid.

Reply:
I’ll debug it. But there is strange behavior when running application, span sended once to apm server.

Reply:
Maybe one of the spans is related to some lifecycle thing that is not related to the request (although I can't see why `render` be invoked in such route)? When you debug, check if one of these spans (the fault span) doesn't have the transaction set as parent.

Reply:
I found the reason, render method called twice from TestDispatcherServlet, too:  .and(hasSuperType(named("org.springframework.web.servlet.DispatcherServlet"))).

Reply:
maybe, instead of `hasSuperType(named("org.springframework.web.servlet.DispatcherServlet"))`, just do `named("org.springframework.web.servlet.DispatcherServlet")`

Reply:
Great, so we have two options:

1. If we expect that `org.springframework.web.servlet.DispatcherServlet#render` is always invoked even when overridden (meaning - overrides always invoke `super#render`) and if we think it is good enough to trace only it (i.e. the duration measurement is good enough etc.) - then just change the matching code above so that only this exact method is traced. That's the easiest, most performant and cleanest solution.
2. Otherwise, we need to guard from nested spans, see https://github.com/elastic/apm-agent-java/blob/5de143707981b2db950120d1519c4327c4aad2da/apm-agent-plugins/apm-httpclient-core/src/main/java/co/elastic/apm/agent/http/client/HttpClientHelper.java#L54 as example. 
In this case, there may or may not be a preference to the outer or inner invocation, so needs to be tested.